### ISSUE TITLE
🦝  NixOS 22.11 - Zero Hydra Failures

### ISSUE THREAD

**Issue by mweinelt** [NixOS/nixpkgs#199919]
Hi, we are Janne Heß & Martin Weinelt, the release managers for NixOS 22.11 :raccoon: ("Raccoon").

Today we want to invite everyone to participate in the Zero Hydra Failures Project, wherein we prepare the package set for the upcoming release, up until its public release at the end of november.

There are only two more upcoming dates that we'd like to mention in that context:

 - 2022-11-21: Branch-off
 - 2022-11-30: 22.11 Release

The complete timeline can be found in the

 - #193585.

## The mission
Every time we plan to do a release, we take time to stabilize the master branch and later on the release branch. Our goal here is to reduce the number of failing jobs on the `nixpkgs:trunk` `nixos:trunk-combined` jobsets as much as possible before branch-off. We call this the "Zero Hydra Failure" campaign.

Besides aiming for zero failed jobs, we also strive to again provide all packages that were available in the previous release.

Changes need to always target the `master` branch. Take note that the branch-off will occur on 2022-11-21, after which ZHF changes will need to be tagged with the `backport: release-22.11` label to land in the stable release.

## Jobsets

The relevant jobsets to check for failing jobs are:

- [nixos:trunk-combined](https://hydra.nixos.org/jobset/nixos/trunk-combined) (Linux packages & NixOS tests)
- [nixpkgs:trunk](https://hydra.nixos.org/jobset/nixpkgs/trunk) (Linux & Darwin packages)

#### Eval reports
Evaluation reports provide a structural overview of the most impactful failing builds. They originated at https://github.com/nix-community/nix-review-tools and were automated over at https://github.com/malob/nix-review-tools-reports.

1. Navigate to https://malob.github.io/nix-review-tools-reports/
2. Open the relevant jobset
3. Browse the latest reports for build failures
![](https://md.darmstadt.ccc.de/uploads/upload_7bab0fdf57d45370e48ba62b43dfbc13.png)
4. Follow the links to the build failure on hydra

#### ZERO Hydra Failures

The platform automatically crawls Hydra and lists packages by maintainer and lists the most important dependencies (failing packages with the most dependants). It also graphs the general trend per platform.

1. Navigate to https://zh.fail

For the record, we started ZHF here:

```
Latest Linux evaluation (completely built):	1784776 on 2022-11-05 07:52:49 (UTC)
Latest Darwin evaluation (completely built):	1784782 on 2022-11-05 09:24:59 (UTC)
Failing builds on aarch64-linux:	1706
Failing builds on i686-linux:	262
Failing builds on x86_64-darwin:	1057
Failing builds on x86_64-linux:	2695
Total failed builds	5720
```

#### Check on packages you maintain

1. Clone nixpkgs and checkout the `master` branch
2. Run 
   ```
   nix-build maintainers/scripts/build.nix --argstr maintainer <name>
   ```
#### Hydra
Hydra is nixpkgs CI platform, where all active branches are built and pushed into the cache, after which channels can originate from its build results.

1. Open the [nixpkgs:trunk](https://hydra.nixos.org/jobset/nixpkgs/trunk) jobset
2. Select the latest evaluation
![](https://md.darmstadt.ccc.de/uploads/upload_968849d97f3c7aea84d173d36f360eac.png)
4. Directly failing jobs are marked with a red cross, while transitively failing ones are greyed out. 
![](https://md.darmstadt.ccc.de/uploads/upload_fb4efdad788e6f904fee43b6ebb40701.png)
5. Use the search form to scope the package list to things relevant to you and that you can test.
![](https://md.darmstadt.ccc.de/uploads/upload_885e122fe1c5ee6dda9b8b3eeffae267.png)

### Submit fixes

6. Search through PRs to make sure noone provided a fix yet. If there is one, please take the time and help review the change.

7. If there is no open PR, troubleshoot why it's failing and fix it.

8. Pull Request the fix against the `master` branch and wait potential review & change requests
    - If your PR causes more than ~500 rebuilds, it is generally preferred to target `staging` to avoid compute churn for users on `master`.

### Backporting

**After 2022-11-21**

9. Apply the relevant backport label to land the fix in the release branch
    - Changes to `master` get backported into `release-22.11`
    - Changes to `staging` get backported into `staging-22.11`

10. If the backport action fails, follow the [manual backporting steps](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#backporting-changes). Make sure to use `git cherry-pick -x <rev>` on all commits intended for backport.
    - @jonringer created a [video covering the backport process](https://www.youtube.com/watch?v=4Zb3GpIc6vk).

---

Always link back to this issue by mentioning the issue number in the description of your pull request:
```
ZHF: #199919
```

If your PR receives no reviews or does not get merged, feel free to
- add the `0.kind: ZHF Fixes` label, so people can better browse these fixes
- request or mention @NixOS/nixos-release-managers on the PR

## Broken packages
Everything we cannot fix in time will need to be marked broken on the respective platforms, so that Hydra will not retry builds over and over, thereby wasting compute resources.

Set `meta.broken` and add a reference and/or explanation, like this:

```nix
meta = {
  # ref to issue/explanation
  broken = stdenv.isDarwin; # only broken on darwin
  # broken = true; # broken on all platforms.
};
```

## Closing
This is a great way to help NixOS, and it is a great time for new contributors to start their nixpkgs adventure. :partying_face: 

As with the [feature freeze issue](https://github.com/NixOS/nixpkgs/issues/194208), please keep the discussion here to a minimum so we don't ping all maintainers (although relevant comments can of course be added here if they are directly ZHF-related) and ping one of the release managers (@mweinelt, @dasJ) in the respective issues.

**Comment by mweinelt** [comment]
Reserved post.

cc @NixOS/nixpkgs-committers @NixOS/nixpkgs-maintainers

**Comment by StarGate01** [comment]
@

**Comment by sg2002** [comment]
Is it ok if I open a separate hub issue for this ZHF concerning the i686 platform? I'm willing to review at least some of the current failures and then either try to get them working, or in the case of some stuff that nobody is realistically willing to maintain, suggesting the Hydra jobs be removed.

**Comment by mweinelt** [comment]
Sure, feel free to put work into that, if you want. But as mentioned above we won't hold back the release for that.

**Comment by mweinelt** [comment]
The branch-off just happened. Remember to add the backport labels for either `release-22.11` or `staging-22.11` for changes that should be included in the release.

**Comment by sikmir** [comment]
Almost all Go packages are broken on darwin: `Get "https://proxy.golang.org/...": x509: OSStatus -26276`. Any ideas what's the reason?

**Comment by winterqt** [comment]
Bump your Nix version, or disable the sandbox temporarily.

**Comment by necrophcodr** [comment]
> Almost all Go packages are broken on darwin: `Get "https://proxy.golang.org/...": x509: OSStatus -26276`. Any ideas what's the reason?

The error seems to suggest a certificate issue. The error code doesn't appear super descriptive either: https://www.osstatus.com/search/results?platform=all&framework=all&search=26276

**Comment by sikmir** [comment]
> Bump your Nix version, or disable the sandbox temporarily.

Sandbox is already disabled:
```
$ grep "sandbox =" /etc/nix/nix.conf
sandbox = false
$ nix --version
nix (Nix) 2.10.1
```

**Comment by winterqt** [comment]
This shouldn't be occurring then.

Either way, can you try bumping your Nix version to the latest one, and report back if it works? Thank you!

**Comment by sikmir** [comment]
> This shouldn't be occurring then.
> 
> Either way, can you try bumping your Nix version to the latest one, and report back if it works? Thank you!

No, the same problem with Nix 2.11.0.

**Comment by winterqt** [comment]
@sikmir I have no idea, then -- I fixed this in Nix 2.11 (see #182889).

Can you please provide a full verbose log (`-vvvvv` should do the trick)? Alternatively, to avoid spamming this thread, you can reach out to me on Matrix (`@winterqt:nixos.dev`) or IRC (`winter` on Libera and OFTC) so we can do less noisy back and forth debugging.

Thank you!
